src/blockencodings.cpp

```diff
@@ -125,20 +128,32 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
                     txn_available[idit->second] = vTxHashes[i].second->GetSharedTx();
                     have_txn[idit->second]  = true;
                     mempool_count++;
+                    missing_txs--;
                 } else {
                     // If we find two mempool txn that match the short id, just request it.
                     // This should be rare enough that the extra bandwidth doesn't matter,
                     // but eating a round-trip due to FillBlock failure would be annoying
                     if (txn_available[idit->second]) {
                         txn_available[idit->second].reset();
-                        mempool_count--;
+                        shortid_collision = true;
```

Comment:
Because there's practically no performance impact of having a shortid_collision bool, and it makes the logic easier to understand.